packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,417 @@
+                // indicating a failure to replay. In that case, we must not
+                // settle the outcomeVow, since the outcome vow only represents
+                // the settled result of the async function itself.
+                // Fortunately, `panic` resets the bijection, again resulting
+                // in the `guestResultP` being absent from the bijection,
+                // so this leave the outcome vow unsettled, as it must.
+                if (bijection.hasGuest(guestResultP)) {
+                  outcomeKit.resolver.reject(membrane.guestToHost(guestReason));
+                  admin.done();
+                }
+              },
+            );
+          },
+          wake() {
+            const { state, facets } = this;
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            if (failures.has(flow)) {
+              throw failures.get(flow);
+            }
+            if (state.isDone) {
+              return;
+            }
+            if (eph.membrane) {
+              eph.membrane.wake();
+            } else {
+              flow.restart();
+            }
+          },
+          getOutcome() {
+            const { state, facets } = this;
+            const { outcomeKit } = state;
+            const { flow } = facets;
+
+            flow.wake();
+            return outcomeKit.vow;
+          },
+          dump() {
+            const { state } = this;
+            const { log } = state;
+
+            return log.dump();
+          },
+          getOptFatalProblem() {
+            const { facets } = this;
+            const { flow } = facets;
+
+            return failures.has(flow) ? failures.get(flow) : undefined;
+          },
+        },
+        admin: {
+          reset() {
+            const { state, facets } = this;
+            const { bijection, log } = state;
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            if (failures.has(flow)) {
+              failures.delete(flow);
+            }
+            if (eagerWakers.has(flow)) {
+              // For now, once an eagerWaker, always an eagerWaker
+              // eagerWakers.delete(flow);
+            }
+            if (eph.membrane) {
+              eph.membrane.stop();
+            }
+            tmp.resetFor(flow);
+            log.reset();
+            bijection.reset();
+
+            state.isDone = false;
+          },
+          done() {
+            const { state, facets } = this;
+            const { log } = state;
+            const { admin } = facets;
+
+            admin.reset();
+            state.isDone = true;
+            log.dispose();
+          },
+          panic(fatalProblem) {
+            const { state, facets } = this;
+            const { bijection, log } = state;
+            const { flow } = facets;
+            const eph = tmp.for(flow);
+
+            if (failures.has(flow)) {
+              failures.set(flow, fatalProblem);
```

Comment:
Since exceptions can be thrown internally due to asyncFlow infrastructure failures, I don't want to rule out that we get an infrastructure failure while trying to handle an infrastructure failure. I also don't want to drop an informative diagnostic in favor of a diagnostic about why we couldn't report a diagnostic. However, the code you comment on does not uphold this principle, since it is the first failure likely to be the more important one.

So I changed it to keep the first error as primary but annotate it with the new error. Done.